Player.cpp

```diff
@@ -0,0 +1,33 @@
+#include "Player.h"
+
+Player::Player(const std::string* characterAttributes)
+	: Character(characterAttributes)
+{
+	XP = 0;
+}
+
+void Player::levelup(float levelupXP)
+{
+	while (levelupXP >= 100)
+	{
+		levelupXP -= 100;
+		HP = maxHP = static_cast<float>(round(maxHP * 1.1));
+		DMG = static_cast<float>(DMG * 1.1);
+	}
+}
+
+void Player::gainXP(float damagePoints)
+{
+	levelup(XP - int(XP / 100) * 100 + damagePoints);
+	XP += damagePoints;
+}
+
+Player Player::PlayerFromFile(const std::string &fileName)
+{
+	return Player(parseUnit(fileName));
+}
+
+void Player::doHit(Character& victim)
+{
+	this->gainXP(victim.gotHit(this));

```

Comment:
Nem kéne ezt függővé tenni a `Character::gotHit` visszatérési értékétől. Inkább külön legyen meghívva a `gainXP` a damage-gel.

Reply:
Azért is döntöttem így, hogy ne kelljen kétszer kiszámolni azt, hogy a támadó teljes sebzése, vagy az áldozat maradék hp-ja a kapott XP. Na meg így egy függvényt kell hívni egy támadásra.   
Vagy csak az úgy legyen, hogy a doHit() függvényen belül ne egyenesen a paraméter legyen a gotHit(), hanem előbb egy <float> experiencePoints (vagy inflictedDamage) változót hozzak létre neki, és aztán
```
this->gainXP(experiencePoints)
```
?

Reply:
**A**   
A gotHit() private lesz visszatérési érték nélkül. A gainXP()-t virtuálisként létre kell hozni a Character osztályban, hogy meg lehessen arra is hívni, csak akkor ne utaljon a csak az alosztályban létező XP tagváltozóra. Ebben az esetben a doHit() feleslegessé válik. A Player::gainXP() overloadban ki kellene számolni újra a bevitt HP-t, mint potenciális XP-t a növelés előtt. Ehhez ugye nem float bemenet kell, hanem az áldozatot kapja majd referenciaként a meghíváskor. Aztán a main()-ben a doHit() helyett egy gainXP(), majd egy gotHit() kerül meghívásra mindig a másikra referálva.
     
**B**  
```
void Player::doHit(Character& victim)
{
        float inflictedDamage = victim.gotHit(this);
	this->gainXP(inflictedDamage);
```   
    
**C**  
Valami, amire nem gondolok egyáltalán.

Reply:
Esetleg még annyit lehet tenni, hogy a tényleges sebzés kiszámolását kiszervezem egy Character osztály beli protected  függvénybe. Ezt a támadóra kell meghívni, mert a gotDMG() protected, de a gotHP() nem (fight során csak a HP-t kell kiíratni). Így legalább a kód nem lesz redundáns, de a számítást ugyanúgy kétszer végzi el a fight minden turnusában.